src/include/migraphx/op/unique.hpp

```diff
@@ -0,0 +1,340 @@
+// ------------------------------------------------------------------
+// When an axis is specified the subarrays indexed by the axis are
+// sorted. This is done by making the specified axis the first
+// dimension of the array (move the axis to the first dimension to
+// keep the order of the other axes) and then flattening the subarrays
+// in C order. The flattened subarrays are then viewed as a structured
+// type with each element given a label, with the effect that we end
+// up with a 1-D array of structured types that can be treated in the
+// same way as any other 1-D array. The result is that the flattened
+// subarrays are sorted in lexicographic order starting with the first
+// element.
+// ------------------------------------------------------------------
+
+struct unique
+{
+    // CASE SORTED:
+    //
+    // To process into a sorted unique series of elements/chunks:
+    // Chunk size == 1 means a simple element; >1 means a flat representation.
+    // Steps: first go through the input elements/chunks for uniqueness.
+    // At the end of this processing, per the sorted sequence of unique elements:
+    // update/create data structures: y, y_indices, x_rev_indices, y_count
+    //
+    // INPUT x: [2, 1, 1, 3, 4, 3], attr_sorted = 1;
+
+    // OUTPUT(s): indices..
+    // y_indices: [1, 0, 3, 4]  --- first incidence, in terms of index in sequence x
+    // x_rev_indices: [1, 0, 0, 2, 3, 2] --- x seen in terms of indices of unique sequence y
+    // y_count: [2, 1, 2, 1] -- count at each y_index. sum = len(x)
+
+    // NOTE: y [1, 2, 3, 4]   --- the unique output is constructed from x[y_indices[...]]
+
+    template <class T>
+    auto sorted_uniq_indices(const T& input_data, size_t chunk_sz) const
+    {
+        struct y_info
+        {
+            size_t y_idx;
+            size_t x_idx;
+            size_t ct = 0;
+        };
+
+        auto idx_less_fn = [&](auto idx1, auto idx2) {
+            // for efficiency, loop is bypassed for single elements. ct = 1
+            for(size_t ct = chunk_sz; ct > 1; idx1++, idx2++, ct--)
+            {
+                if(input_data[idx1] < input_data[idx2])
+                    return true;
+                else if(input_data[idx1] > input_data[idx2])
+                    return false;
+            }
+            return input_data[idx1] < input_data[idx2];
+        };
+        std::map<size_t, y_info, decltype(idx_less_fn)> uniq_val_map(idx_less_fn);
+
+        std::vector<std::size_t> x_rev_indices;
+
+        // go through all the elements and find the unique elements..
+        size_t count_x = input_data.size();
+        for(size_t f_idx = 0, x_idx = 0; f_idx < count_x; f_idx += chunk_sz, x_idx++)
+        {
+            y_info entry          = {.y_idx = uniq_val_map.size(), .x_idx = x_idx};
+            auto [itr, added_new] = uniq_val_map.insert({f_idx, entry});
+            itr->second.ct++;
+            x_rev_indices.push_back(itr->second.y_idx);
+        }
```

Comment:
I think it would be much simpler to write this as:

```cpp
for(auto x_idx:range(input_data.size() / chunk_sz))
{
    size_t f_idx = x_idx * chunk_sz;
    y_info entry          = {.y_idx = uniq_val_map.size(), .x_idx = x_idx};
    auto [itr, added_new] = uniq_val_map.insert({f_idx, entry});
    itr->second.ct++;
    x_rev_indices.push_back(itr->second.y_idx);
}
```

Reply:
This is still not resolved. Can you write it similiar to what I've shown? Using multiple statements in the `for` loop makes it harder to decipher whats happening. Using a range-based for loop makes it much simpler.

Reply:
@pfultz2  Thanks for your suggestion. Both -- the current style and your suggested style -- are efficient ways to implement this small for-loop. 

Code gets complicated with number of lines in a loop, and its readability becomes an issue then. These 4 lines of for-loop (or 5 in your case) are really way too small for anybody to get confused, hence my suggestion is to just leave it as it is. Thanks.

Reply:
Its much simpler to have single statements in the for loop. I will add a cppcheck rule for this in the future.

Reply:
Thanks. I suggest we instead check all loops for their size -- and more importantly for their complexity, when there are confusing branching and loop-size changes within that loop: the sentiment expressed by Sean Parent, that we both would readily agree to, I am sure.

Reply:
> I suggest we instead check all loops for their size

We do have size checks for functions with [`readability-function-size`](https://clang.llvm.org/extra/clang-tidy/checks/readability/function-size.html) and would like to enable [`readability-function-cognitive-complexity`](https://clang.llvm.org/extra/clang-tidy/checks/readability/function-cognitive-complexity.html) but there is some code that doesnt pass this check.

However, these are just checking functions, it would nice to have these for loops as well. And even more so we want to check for multiple statements so develops dont hack it to work by putting more statements in the `for` to avoid complexity in the body.